### PR TITLE
fix(hooks): [use-focus-controller] repeatedly trigger focus when clicked

### LINKED ISSUE #16959: [Component] [input] focus事件里面执行blur，会再次出发focus事件
Bug Type: **`Component`**

## Environment

- Vue Version: `3.4.3`
- Element Plus Version: `2.7.3`
- Browser / OS: `Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.3`
- Build Tool: `Vite`

### Related Component

- `el-input`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbiAgPGVsLWlucHV0IHJlZj1cImlucHV0UmVmXCIgdi1tb2RlbD1cImlucHV0XCIgc3R5bGU9XCJ3aWR0aDogMjQwcHhcIiBwbGFjZWhvbGRlcj1cIlBsZWFzZSBpbnB1dFwiIEBmb2N1cz1cImhGb2N1c1wiIC8+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IGxhbmc9XCJ0c1wiIHNldHVwPlxuaW1wb3J0IHsgcmVmIH0gZnJvbSAndnVlJ1xuY29uc3QgaW5wdXQgPSByZWYoJycpXG5jb25zdCBpbnB1dFJlZiA9IHJlZjxIVE1MSW5wdXRFbGVtZW50PigpXG5jb25zdCBoRm9jdXMgPSAoKSA9PiB7XG4gIGlucHV0UmVmLnZhbHVlPy5ibHVyKClcbiAgY29uc29sZS5sb2coJ2ZvY3VzJylcblxufVxuPC9zY3JpcHQ+IiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge31cbn0iLCJ0c2NvbmZpZy5qc29uIjoie1xuICBcImNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogXCJFU05leHRcIixcbiAgICBcImpzeFwiOiBcInByZXNlcnZlXCIsXG4gICAgXCJtb2R1bGVcIjogXCJFU05leHRcIixcbiAgICBcIm1vZHVsZVJlc29sdXRpb25cIjogXCJCdW5kbGVyXCIsXG4gICAgXCJ0eXBlc1wiOiBbXCJlbGVtZW50LXBsdXMvZ2xvYmFsLmQudHNcIl0sXG4gICAgXCJhbGxvd0ltcG9ydGluZ1RzRXh0ZW5zaW9uc1wiOiB0cnVlLFxuICAgIFwiYWxsb3dKc1wiOiB0cnVlLFxuICAgIFwiY2hlY2tKc1wiOiB0cnVlXG4gIH0sXG4gIFwidnVlQ29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiAzLjNcbiAgfVxufVxuIiwiX28iOnt9fQ==)

### Steps to reproduce

点击输入框，查看console

### What is Expected?

失去焦点不再触发focus

### What is actually happening?

会触发focus，执行两次

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by tolking**
closed #16959, closed #16961

Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Review by warmthsea — COMMENTED** [review]
I thought so at first, but then I remembered that I used to use input to trigger the outer click. 🤔

**Review by tolking — COMMENTED** [review]
> I thought so at first, but then I remembered that I used to use input to trigger the outer click. 🤔 

Can you provide a demo

**Comment by warmthsea** [comment]
> > I thought so at first, but then I remembered that I used to use input to trigger the outer click. 🤔
> 
> Can you provide a demo

What I mean is that if add `@click.stop`, similar usage will not work. This is a break change.

[demo](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbiAgPGRpdiBjbGFzcz1cImgtMjAgcHgtMTAgYmctYmx1ZS0xMDAgZmxleCBpdGVtcy1jZW50ZXJcIlxuICAgIEBjbGljaz1cInN0b3JlVGFnID0gZmFsc2VcIlxuICA+XG4gICAgaGVhZGVyIGNsaWNrIG1lXG4gIDwvZGl2PlxuICA8ZGl2IGNsYXNzPVwiaC05NiBiZy1yZWQtNTAgZmxleCBpdGVtcy1jZW50ZXIganVzdGlmeS1jZW50ZXJcIlxuICAgIEBjbGljaz1cInN0b3JlVGFnID0gdHJ1ZVwiXG4gID5cbiAgICA8cCBjbGFzcz1cIm1yLThcIj57eyBzdG9yZVRhZyB9fTwvcD5cbiAgICA8ZGl2PlxuICAgICAgPGVsLWlucHV0XG4gICAgICAgdi1tb2RlbD1cImlucHV0XCIgXG4gICAgICAgcGxhY2Vob2xkZXI9XCJDbGljayBtZVwiXG4gICAgICAvPlxuICAgIDwvZGl2PlxuICA8L2Rpdj5cbjwvdGVtcGxhdGU+XG5cbjxzY3JpcHQgbGFuZz1cInRzXCIgc2V0dXA+XG5pbXBvcnQgeyByZWYgfSBmcm9tICd2dWUnXG5jb25zdCBpbnB1dCA9IHJlZignJylcblxuLy8gVGhpcyB2YWx1ZSBjYW4gYmUgc3RvcmVkIGluIGFuIGV4dGVybmFsIHN0b3JlXG5jb25zdCBzdG9yZVRhZyA9IHJlZihmYWxzZSlcbjwvc2NyaXB0PlxuIiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge31cbn0iLCJ0c2NvbmZpZy5qc29uIjoie1xuICBcImNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogXCJFU05leHRcIixcbiAgICBcImpzeFwiOiBcInByZXNlcnZlXCIsXG4gICAgXCJtb2R1bGVcIjogXCJFU05leHRcIixcbiAgICBcIm1vZHVsZVJlc29sdXRpb25cIjogXCJCdW5kbGVyXCIsXG4gICAgXCJ0eXBlc1wiOiBbXCJlbGVtZW50LXBsdXMvZ2xvYmFsLmQudHNcIl0sXG4gICAgXCJhbGxvd0ltcG9ydGluZ1RzRXh0ZW5zaW9uc1wiOiB0cnVlLFxuICAgIFwiYWxsb3dKc1wiOiB0cnVlLFxuICAgIFwiY2hlY2tKc1wiOiB0cnVlXG4gIH0sXG4gIFwidnVlQ29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiAzLjNcbiAgfVxufVxuIiwiX28iOnt9fQ==)

File: packages/components/input/src/input.vue [hunk 1]
```diff
@@ -129,6 +130,7 @@
         @blur="handleBlur"
         @change="handleChange"
         @keydown="handleKeydown"
+        @click.stop
```
Comment (btea) [inline packages/components/input/src/input.vue:133]: The focus and blur events of textarea are bound directly, not by proxy, so there seems to be no problem, and no modification is required.

File: packages/hooks/use-focus-controller/index.ts [hunk 2]
```diff
@@ -56,10 +56,11 @@ export function useFocusController<T extends { focus: () => void }>(
     afterBlur?.()
   }
 
-  const handleClick = () => {
+  const handleClick = (event: Event) => {
     if (
       (wrapperRef.value?.contains(document.activeElement) &&
         wrapperRef.value !== document.activeElement) ||
+      isFocusable(event.target as HTMLElement) ||
       disabled.value
```
Comment (Dsaquel) [inline packages/hooks/use-focus-controller/index.ts:64]: I feel like the condition is abnormal : if the target element is focusable, then maybe we want to focus this element ?

Maybe it is better to let the browser event handle the focus automatically, i tried to remove the click listener and only one test failing:
https://github.com/element-plus/element-plus/blob/d3dae80b70d096c32e922f969986e7a0aa4e8ad3/packages/components/time-picker/__tests__/time-picker.test.tsx#L873

I think it is maybe better to remove the click listener here and handle edge cases in their places.

Comment (tolking) [reply packages/hooks/use-focus-controller/index.ts:64]: Because the trigger order of the focusable elements is: focus -> click -> focus (~~emit~~)

The elements that cannot be focused is: click -> focus

After change, the focusable elements is: focus -> click

---

Deleting the click listener is not conducive to focus management, as the clear button will cause the component to lose focus

Comment (Dsaquel) [reply packages/hooks/use-focus-controller/index.ts:64]: >The elements that cannot be focused is: click -> focus
>
I was aware that the clear icon can be focused in this case but i test it and it is not so that why the fallback to the input focus.

Ok i think i got it thanks for the explanation.

**Review by Dsaquel — APPROVED** [review]

**Comment by btea** [comment]
Perhaps we can add some relevant test cases.

**Comment by tolking** [comment]
> Perhaps we can add some relevant test cases.

I have tried writing tests, but now the click events in the tests cannot trigger the focus events.

**Review by btea — APPROVED** [review]

**Review by warmthsea — APPROVED** [review]
